arwen/host/asyncCall.go

```diff
@@ -0,0 +1,150 @@
+package host
+
+import (
+	"math/big"
+
+	arwen "github.com/ElrondNetwork/arwen-wasm-vm/arwen"
+	"github.com/ElrondNetwork/arwen-wasm-vm/wasmer"
+	vmcommon "github.com/ElrondNetwork/elrond-vm-common"
+)
+
+func (host *vmHost) handleAsyncCallBreakpoint(result wasmer.Value) error {
+	runtime := host.Runtime()
+	runtime.SetRuntimeBreakpointValue(arwen.BreakpointNone)
+
+	// TODO also determine whether caller and callee are in the same Shard (based
+	// on address?), by account addresses - this would make the empty SC code an
+	// unrecoverable error, so returning nil here will not be appropriate anymore.
+	if !host.canExecuteSynchronously() {
+		return nil
+	}
+
+	// Start calling the destination SC, synchronously.
+	destinationCallInput, err := host.createDestinationContractCallInput()
+	if err != nil {
+		return err
+	}
+
+	destinationVMOutput, err := host.ExecuteOnDestContext(destinationCallInput)
```

Comment:
it is pretty easy to make these few calls in a recursive way. :)

Reply:
Yes, async call reentrance will be much easier to implement after the last refactoring.